apis/r/R/SOMADataFrame.R

```diff
@@ -126,52 +126,55 @@ SOMADataFrame <- R6::R6Class(
-    #' @param value_filter A string containing a logical expression that is used
+    #' @param ids Optional indices specifying the rows to read.
+    #' @param column_names Optional character vector of column names to return.
+    #' @param value_filter Optional string containing a logical expression that is used
     #' to filter the returned values. See [`tiledb::parse_query_condition`] for
     #' more information.
-    #' @param result_order Order of read results. This can be one of either
+    #' @param result_order Optional order of read results. This can be one of either
     #' `"ROW_MAJOR, `"COL_MAJOR"`, `"GLOBAL_ORDER"`, or `"UNORDERED"`.
+    #' @param log_level Optional logging level with default value of `"warn"`.
     #' @return An [`arrow::Table`].
-    read = function(
-      ids = NULL,
-      column_names = NULL,
-      value_filter = NULL,
-      result_order = "UNORDERED"
-    ) {
-      on.exit(private$close())
-      private$open("READ")
+    read = function(ids = NULL,
+                    column_names = NULL,
+                    value_filter = NULL,
+                    result_order = "UNORDERED",
+                    log_level = "warn") {
 
-      arr <- self$object
+      result_order <- match_query_layout(result_order)
 
-      # select columns
-      if (!is.null(column_names)) {
-        stopifnot(
-          "'column_names' must only contain non-index columns" =
-            all(!column_names %in% self$dimnames())
-        )
-        tiledb::attrs(arr) <- column_names
-      }
+      uri <- self$uri
+      arr <- self$object                 # need array (schema) to properly parse query condition
 
-      # select ranges
-      if (!is.null(ids)) {
-        tiledb::selected_ranges(arr) <- list(cbind(ids, ids))
-      }
+      # check columns
+      stopifnot("'column_names' must only contain valid dimension or attribute columns" =
+                    is.null(column_names) ||
+                    all(column_names %in% c(self$dimnames(), self$attrnames())))
 
-      # filter
+      # check and parse value filter
+      stopifnot("'value_filter' must be a single argument" =
+                    is.null(value_filter) || is_scalar_character(value_filter))
       if (!is.null(value_filter)) {
-        stopifnot(is_scalar_character(value_filter))
-        tiledb::query_condition(arr) <- do.call(
-          what = tiledb::parse_query_condition,
-          args = list(expr = str2lang(value_filter), ta = self$object)
-        )
+          parsed <- do.call(what = tiledb::parse_query_condition,
+                            args = list(expr = str2lang(value_filter), ta = arr))
+          value_filter <- parsed@ptr
+      }
+      # ensure ids is a (named) list
+      if (!is.null(ids)) {
+          if (!is.list(ids))
+              ids <- list(ids)
+          if (is.null(names(ids)))
+              names(ids) <- self$dimnames() # TODO: ensure dimensionality
```

Comment:
I think we should require users to name lists passed to `ids` b/c we can't assume the order (or number) of elements matches the array's dimensions (unless it's a 1D array).

Reply:
Now done.